pulp_rpm/app/metadata_parsing.py

```diff
@@ -0,0 +1,130 @@
+            other_root_element = next(other_xml_iterator)[1]
+        # I know. This is a terrible misuse of SyntaxError. Don't blame the messenger.
+        except SyntaxError:
+            log.error("failed to parse XML metadata file")
+            raise
+
+        while True:
+            for event, filelists_element in filelists_xml_iterator:
+                # if we're not at a fully parsed package element, keep going
+                if event != "end":
+                    continue
+                # make this work whether the file has namespace as part of the tag or not
+                if not (
+                    filelists_element.tag == "package"
+                    or re.sub(NS_STRIP_RE, "", filelists_element.tag) == "package"
+                ):
+                    continue
+
+                break
+
+            for event, other_element in other_xml_iterator:
+                # if we're not at a fully parsed package element, keep going
+                if event != "end":
+                    continue
+                # make this work whether the file has namespace as part of the tag or not
+                if not (
+                    other_element.tag == "package"
+                    or re.sub(NS_STRIP_RE, "", other_element.tag) == "package"
+                ):
+                    continue
+
+                break
+
+            filelists_root_element.clear()  # clear all previously parsed ancestors of the root
+            other_root_element.clear()
+
+            (filelists_pkgid, files) = process_filelists_package_element(filelists_element)
+            (other_pkgid, changelogs) = process_other_package_element(other_element)
+
+            assert (
+                filelists_pkgid == other_pkgid
+            ), "Package key for filelists.xml and other.xml do not match"
+
+            yield filelists_pkgid, files, changelogs
+
+
+def process_filelists_package_element(element):
+    """Process one element from the filelists.xml file and return its parsed data.
+
+    :param element: object representing an XML block for one package's file list
+    :type  element: xml.etree.ElementTree.Element
+
+    :return:    unit key and dictionary containing keys "file" and "dir" where
+                values are full filesystem paths as strings.
+    :rtype:     tuple(dict, dict)
+    """
+    pkgid = element.attrib["pkgid"]
+
+    files = []
+    for element in element.findall("file"):
+        basename, filename = os.path.split(element.text)
+        ftype = element.attrib.get("type")
+
+        files.append((ftype, basename, filename))
+
+    return pkgid, files
+
+
+def process_other_package_element(element):
+    """Parse package element from other.xml."""
+    pkgid = element.attrib["pkgid"]
+
+    changelogs = []
+    for element in element.findall("changelog"):
+        author = element.attrib["author"]
+        date = int(element.attrib["date"])
+        text = element.text
+
+        changelogs.append([author, date, text])
```

Comment:
Any reason why it is list of lists, while for files it's list of tuples?
I think list of tuples makes sense but I'm for any consistency.